Fix /simplify Neuron archdir detection

ferdonline opened the pull request:
Instead of relying on build system dependent, we use (parse)
nrnivmodl to get the exact same
This was affecting mac users, as well as the FMT lib

ferdonline committed:
Fix /simplify Neuron archdir detection

Instead of relying on build system dependent, we use (parse)
nrnivmodl to get the exact same
This was affecting mac users, as well as the FMT lib

matz-e commented on var/spack/repos/builtin/packages/neuron/package.py:
```diff
@@ -368,23 +368,10 @@ def archdir(self):
-            else:
-                neuron_arch = ""
-        return neuron_arch
+        return subprocess.check_output(
+            ['awk', '-F=', '$1 == "MODSUBDIR" { print $2; exit; }',
+             str(self.prefix.bin.nrnivmodl)]
+        ).decode().strip()
```
That's a lot simpler :)

matz-e replied:
It's not much more verbose:
```suggestion
        return subprocess.Popen(
            ['awk', '-F=', '$1 == "MODSUBDIR" { print $2; exit; }',
             str(self.prefix.bin.nrnivmodl)],
             stdout=subprocess.PIPE
        ).communicate()[0].decode().strip()
```

ferdonline commented:
CI says: `python_version.py::test_package_module_compatibility FAILED : AssertionError: assert not {(2, 7): `
Do we really want to force py27 compat? 😞

pramodk commented:
> Do we really want to force py27 compat? 😞

I would put in other words - I wouldn't break it if there is no major issue. The reason being, if we start building stack on system where there is only Python2 on login nodes (e.g. Marconi), we can build everything ourselves from scratch.

matz-e commented:
I think ditching Python 2 is a real option with future Spack

pramodk commented:
> I think ditching Python 2 is a real option with future Spack

In that case we assume that Python3 is **always** available where we want to build the software stack. I would defer this switch to Python3.

matz-e commented:
> > I think ditching Python 2 is a real option with future Spack
> 
> In that case we assume that Python3 is **always** available where we want to build the software stack. I would defer this switch to Python3.

This is a discussion upstream. Spack will already pick up an installed Python 3 and use it, I guess it's just a question of time until support for 2 is dropped.

If I dig out my old copy of Ubuntu Warty Warthog, Spack will not run there, either…

pramodk commented:
> This is a discussion upstream.

Exactly! (and IIRC discussions, this will be a conservative choice)

If you upgrade from `Warty Warthog` to `Intrepid Ibex` it should work!

ferdonline commented:
Ohh, it's actually worse than that. It works under 2.7. Oh gosh, 2.6 is like yet another language...
```
---------------------------- Captured stdout call -----------------------------
==> Spack must remain compatible with Python version 2.6
==> These files require version 2.7
```

ferdonline committed:
Fix to spack to allow py2.7 package recipes

Spack recipes have long been using python3 language features so we
should allow py27 code.
With this patch spack still makes a warning on recipes that requires py 2.7,
but will still raise Exception if core spack files dont run in 2.6

ferdonline commented:
My solution might seem a bit harsh, but I'm pretty sure we have recipes that need at least py2.7 to run, so to stop this 2.6 ideology, can we allow our recipes to use py2.7? @pramodk ?

matz-e commented on lib/spack/spack/test/python_version.py:
```diff
@@ -89,14 +89,15 @@ def pyfiles(search_paths, exclude=()):
 
 
-def check_python_versions(files):
+def check_python_versions(files, warn_only_py26=False):
```
Mucking around in core will make future merges harder. I'm actually trying to move away from all the workarounds we have in there.

ferdonline replied:
In the next change they do to this file I do hope they get rid of python 2.6 compatibility enforcement. Do you suggest some lighter changes?

matz-e replied:
Revert the `lib` changes.

ferdonline committed:
Use Popen to stay compat w 2.6 and leave lib alone

ferdonline committed:
pep8 fix

matz-e reviewed (approved):
Thanks!

ferdonline commented:
Tests failed because of a glitch in the testing. I have created a fix in https://bbpcode.epfl.ch/code/#/c/50860/
